src/amo/components/Addon/index.js

```diff
@@ -73,9 +80,21 @@ export class AddonBase extends React.Component {
   }
 
   componentWillMount() {
-    const { addon, dispatch } = this.props;
+    const { addon, dispatch, errorHandler, params } = this.props;
+
+    if (addon) {
+      dispatch(setViewContext(addon.type));
+    } else {
+      dispatch(fetchAddon({ slug: params.slug, errorHandler }));
+    }
+  }
 
-    dispatch(setViewContext(addon.type));
+  componentWillReceiveProps({ addon: newAddon }) {
+    const { addon: oldAddon, dispatch } = this.props;
```

Comment:
Ohh, wow. That's not what I expect, making it a bit hard to read. Maybe it would be nicer to do:

```js
const { dispatch } = this.props;
const addon = this.props.oldAddon;
```

I just find that syntax opposite to what I expect.